### PR TITLE
ENH tokens per-CI service

### LINKED ISSUE #1709: enable a feedstock token per CI service
Per suggestions from @isuruf (and I think @wolfv too a while back? possibly others too!), we should have separate tokens per CI service. This will reduce the number of token rotations we need.

### PR DESCRIPTION
**Pull request by beckermr**
Checklist
* [x] Added a ``news`` entry

closes #1709

### CODE REVIEW

File: conda_smithy/feedstock_tokens.py [hunk 1]
```diff
@@ -22,43 +26,87 @@
...
+    ``feedstock_token_local_path(user, project, provider=provider)``.
 
     This function will fail if the token file already exists.
     """
     # capture stdout, stderr and suppress all exceptions so we don't
     # spill tokens
     failed = False
     err_msg = None
-    with open(os.devnull, "w") as fp, redirect_stdout(fp), redirect_stderr(fp):
+    token = None
```
Comment (beckermr) [inline conda_smithy/feedstock_tokens.py:94]:
```suggestion
```

File: conda_smithy/feedstock_tokens.py [hunk 2]
```diff
@@ -300,12 +360,10 @@ def register_feedstock_token(user, project, token_repo):
             # don't overwrite existing tokens
             # check again since there might be a race condition
```
Comment (beckermr) [inline conda_smithy/feedstock_tokens.py:361]:
```suggestion
            # append the token if needed
```

File: conda_smithy/feedstock_tokens.py [hunk 3]
```diff
@@ -297,15 +357,12 @@ def register_feedstock_token(user, project, token_repo):
...
+            # append the token if needed
             if os.path.exists(token_file):
-                failed = True
-                err_msg = "Token for repo %s/%s already exists!" % (
-                    user,
-                    project,
-                )
-                raise RuntimeError(err_msg)
+                with open(token_file, "r") as fp:
+                    token_data = json.load(fp)
```
Comment (beckermr) [inline conda_smithy/feedstock_tokens.py:363]:
```suggestion
                    token_data = json.load(fp)
                if "tokens" not in token_data:
                    token_data = {"tokens": [token_data]}
```

**Comment by beckermr** [comment]
@conda-forge/core this one is ready for review!

File: tests/test_feedstock_tokens.py [hunk 4]
```diff
@@ -97,45 +179,70 @@ def test_is_valid_feedstock_token_badtoken(
...
-def test_generate_and_write_feedstock_token():
+@pytest.mark.parametrize("ci", [None, "azure"])
+def test_generate_and_write_feedstock_token(ci):
     user = "bar"
     repo = "foo"
 
-    pth = os.path.expanduser("~/.conda-smithy/bar_foo.token")
+    if ci:
+        pth = os.path.expanduser("~/.conda-smithy/bar_foo_%s.token" % ci)
+        opth = os.path.expanduser("~/.conda-smithy/bar_foo.token")
```
[the remark this answers is not in the document]
Comment (beckermr) [reply tests/test_feedstock_tokens.py:202]: No. These are two different temporary token locations on disk.

File: conda_smithy/feedstock_tokens.py [hunk 5]
```diff
@@ -373,127 +442,104 @@ def register_feedstock_token_with_proviers(
...
+            raise e
+        except Exception as e:
+            if "DEBUG_FEEDSTOCK_TOKENS" in os.environ:
+                raise e
+            failed = True
+
+    if failed:
+        raise FeedstockTokenError(
+            (
+                "Registering the feedstock token with proviers for %s/%s failed!"
```
[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/feedstock_tokens.py:537]: Yes, but not in this PR. :)

File: conda_smithy/feedstock_tokens.py [hunk 6]
```diff
@@ -373,127 +442,104 @@ def register_feedstock_token_with_proviers(
...
+                    clobber,
+                    "github_actions",
+                    add_feedstock_token_to_github_actions,
+                )
+        except FeedstockTokenError as e:
+            raise e
+        except Exception as e:
+            if "DEBUG_FEEDSTOCK_TOKENS" in os.environ:
+                raise e
+            failed = True
```
[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/feedstock_tokens.py:532]: Keep trying until the end. We don't hit failures here very often and they usually go away when we retry. Thus it doesn't make a ton of sense to put more effort here.

File: conda_smithy/cli.py [hunk 7]
```diff
@@ -63,7 +62,7 @@ def generate_feedstock_content(target_directory, source_recipe_dir):
         try:
             with open(forge_yml_recipe, "r") as fp:
                 _cfg = yaml.load(fp.read())
-        except:
+        except Exception:
```
[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/cli.py:65]: This was me fixing a bad practice in old code. Not going to change more.

File: conda_smithy/feedstock_tokens.py [hunk 8]
```diff
@@ -22,43 +26,86 @@
...
+    finally:
+        pass
+
 
-        ~/.conda-smithy/{user or org}_{repo}_feedstock.token
+def feedstock_token_local_path(user, project, provider=None):
+    """Return the path locally where the feedstock
+    token is stored.
+    """
+    if provider is None:
```
[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/feedstock_tokens.py:67]: I find that more confusing than the simple if-else structure a beginner would use.

File: conda_smithy/feedstock_tokens.py [hunk 9]
```diff
@@ -22,43 +26,86 @@
...
+
+    pass
+
+
+@contextmanager
+def _secure_io():
+    """context manager that redirects stdout and
+    stderr to /dev/null to avoid spilling tokens"""
+    try:
+        with open(os.devnull, "w") as fp:
```
[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/feedstock_tokens.py:49]: This block was designed to not spill tokens to CI logs, not to absolutely secure in the CI environment.

File: conda_smithy/feedstock_tokens.py [hunk 10]
```diff
@@ -22,43 +26,86 @@
...
+    """Custom exception for sanitzed token errors."""
+
+    pass
+
+
+@contextmanager
+def _secure_io():
+    """context manager that redirects stdout and
+    stderr to /dev/null to avoid spilling tokens"""
+    try:
```
[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/feedstock_tokens.py:48]: I follow the examples in the python docs and things just work, even if a few random pass branches are lying around.

File: conda_smithy/cli.py [hunk 11]
```diff
@@ -806,10 +832,32 @@ def __call__(self, args):
...
-        register_feedstock_token(owner, repo, token_repo)
+        if args.unique_token_per_provider:
+            for ci in [
+                "azure",
+                "travis",
+                "circle",
+                "drone",
+                "github_actions",
+            ]:
+                if getattr(args, ci):
```
[the remark this answers is not in the document]
Comment (beckermr) [reply conda_smithy/cli.py:847]: This won't work. Those ci variables either True or False.

**Comment by beckermr** [comment]
OK @jaimergp! Ready for another look!

**Comment by jakirkham** [comment]
So do we need to regenerate tokens to use this or how does that work?

**Comment by beckermr** [comment]
Yeah there's a whole migration plan. I'll add more notes than the steps I put in the issue. I want to do some testing by hand first.

**Pull request merged**